File: src/services/ai/streaming.ts
import type { AIProvider, LlmRequest, TokenUsage } from '../../types';
import { getProviderConfig } from './providers';

export interface StreamCallbacks {
  onToken: (token: string) => void;
  onComplete: (fullText: string, usage?: TokenUsage) => void;
  onError: (error: string) => void;
}

export async function streamCompletion(
  request: LlmRequest,
  callbacks: StreamCallbacks,
  abortSignal?: AbortSignal
): Promise<void> {
  const provider = request.provider as AIProvider;
  const config = getProviderConfig(provider);

  try {
    if (provider === 'gemini') {
      await streamGemini(request, callbacks, abortSignal);
    } else if (provider === 'anthropic') {
      await streamAnthropic(request, callbacks, abortSignal);
    } else {
      await streamOpenAICompatible(request, config.baseUrl, callbacks, abortSignal);
    }
  } catch (err) {
    if (abortSignal?.aborted) return;
    callbacks.onError(err instanceof Error ? err.message : String(err));
  }
}

async function streamOpenAICompatible(
  request: LlmRequest,
  baseUrl: string,
  callbacks: StreamCallbacks,
  abortSignal?: AbortSignal
): Promise<void> {
  const messages: Array<{role: string; content: string | Array<{type: string; text?: string; image_url?: {url: string}}>}> = [
    { role: 'system', content: request.system_prompt },
  ];

  if (request.images && request.images.length > 0) {
    const content: Array<{type: string; text?: string; image_url?: {url: string}}> = [
      { type: 'text', text: request.user_prompt },
      ...request.images.map((img) => ({
        type: 'image_url' as const,
        image_url: { url: img },
      })),
    ];
    messages.push({ role: 'user', content });
  } else {
    messages.push({ role: 'user', content: request.user_prompt });
  }

  const body = {
    model: request.model,
    messages,
    temperature: request.temperature ?? 0.3,
    max_tokens: request.max_tokens ?? 8192,
    stream: true,
  };

  const headers: Record<string, string> = {
    'Content-Type': 'application/json',
  };

  if (request.api_key) {
    headers['Authorization'] = `Bearer ${request.api_key}`;
  }

  if (request.provider === 'openrouter') {
    headers['HTTP-Referer'] = 'https://zenith-ide.app';
    headers['X-Title'] = 'Zenith IDE';
  }

  const response = await fetch(`${baseUrl}/chat/completions`, {
    method: 'POST',
    headers,
    body: JSON.stringify(body),
    signal: abortSignal,
  });

  if (!response.ok) {
    const errorText = await response.text();
    throw new Error(`API error ${response.status}: ${errorText}`);
  }

  const reader = response.body?.getReader();
  if (!reader) throw new Error('No response body');

  const decoder = new TextDecoder();
  let fullText = '';
  let buffer = '';

  while (true) {
    const { done, value } = await reader.read();
    if (done) break;

    buffer += decoder.decode(value, { stream: true });
    const lines = buffer.split('\n');
    buffer = lines.pop() || '';

    for (const line of lines) {
      const trimmed = line.trim();
      if (!trimmed || trimmed === 'data: [DONE]') continue;
      if (!trimmed.startsWith('data: ')) continue;

      try {
        const json = JSON.parse(trimmed.slice(6));
        const delta = json.choices?.[0]?.delta?.content;
        if (delta) {
          fullText += delta;
          callbacks.onToken(delta);
        }

        if (json.usage) {
          callbacks.onComplete(fullText, {
            prompt_tokens: json.usage.prompt_tokens || 0,
            completion_tokens: json.usage.completion_tokens || 0,
            total_tokens: json.usage.total_tokens || 0,
          });
          return;
        }
      } catch {
        // skip malformed SSE
      }
    }
  }

  callbacks.onComplete(fullText);
}

async function streamGemini(
  request: LlmRequest,
  callbacks: StreamCallbacks,
  abortSignal?: AbortSignal
): Promise<void> {
  const url = `https://generativelanguage.googleapis.com/v1beta/models/${request.model}:streamGenerateContent?alt=sse&key=${request.api_key}`;

  const parts: Array<{text?: string; inlineData?: {mimeType: string; data: string}}> = [
    { text: request.user_prompt },
  ];

  if (request.images) {
    for (const img of request.images) {
      const match = img.match(/^data:(.+?);base64,(.+)$/);
      if (match) {
        parts.push({
          inlineData: { mimeType: match[1], data: match[2] },
        });
      }
    }
  }

  const body = {
    system_instruction: { parts: [{ text: request.system_prompt }] },
    contents: [{ role: 'user', parts }],
    generationConfig: {
      temperature: request.temperature ?? 0.3,
      maxOutputTokens: request.max_tokens ?? 8192,
    },
  };

  const response = await fetch(url, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
    signal: abortSignal,
  });

  if (!response.ok) {
    const errorText = await response.text();
    throw new Error(`Gemini API error ${response.status}: ${errorText}`);
  }

  const reader = response.body?.getReader();
  if (!reader) throw new Error('No response body');

  const decoder = new TextDecoder();
  let fullText = '';
  let buffer = '';

  while (true) {
    const { done, value } = await reader.read();
    if (done) break;

    buffer += decoder.decode(value, { stream: true });
    const lines = buffer.split('\n');
    buffer = lines.pop() || '';

    for (const line of lines) {
      const trimmed = line.trim();
      if (!trimmed.startsWith('data: ')) continue;

      try {
        const json = JSON.parse(trimmed.slice(6));
        const text = json.candidates?.[0]?.content?.parts?.[0]?.text;
        if (text) {
          fullText += text;
          callbacks.onToken(text);
        }
      } catch {
        // skip
      }
    }
  }

  callbacks.onComplete(fullText);
}

async function streamAnthropic(
  request: LlmRequest,
  callbacks: StreamCallbacks,
  abortSignal?: AbortSignal
): Promise<void> {
  const content: Array<{type: string; text?: string; source?: {type: string; media_type: string; data: string}}> = [
    { type: 'text', text: request.user_prompt },
  ];

  if (request.images) {
    for (const img of request.images) {
      const match = img.match(/^data:(.+?);base64,(.+)$/);
      if (match) {
        content.push({
          type: 'image',
          source: { type: 'base64', media_type: match[1], data: match[2] },
        });
      }
    }
  }

  const body = {
    model: request.model,
    system: request.system_prompt,
    messages: [{ role: 'user', content }],
    max_tokens: request.max_tokens ?? 8192,
    temperature: request.temperature ?? 0.3,
    stream: true,
  };

  const response = await fetch('https://api.anthropic.com/v1/messages', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'x-api-key': request.api_key,
      'anthropic-version': '2023-06-01',
      'anthropic-dangerous-direct-browser-access': 'true',
    },
    body: JSON.stringify(body),
    signal: abortSignal,
  });

  if (!response.ok) {
    const errorText = await response.text();
    throw new Error(`Anthropic API error ${response.status}: ${errorText}`);
  }

  const reader = response.body?.getReader();
  if (!reader) throw new Error('No response body');

  const decoder = new TextDecoder();
  let fullText = '';
  let buffer = '';
  let usage: TokenUsage | undefined;

  while (true) {
    const { done, value } = await reader.read();
    if (done) break;

    buffer += decoder.decode(value, { stream: true });
    const lines = buffer.split('\n');
    buffer = lines.pop() || '';

    for (const line of lines) {
      const trimmed = line.trim();
      if (!trimmed.startsWith('data: ')) continue;

      try {
        const json = JSON.parse(trimmed.slice(6));

        if (json.type === 'content_block_delta') {
          const delta = json.delta?.text;
          if (delta) {
            fullText += delta;
            callbacks.onToken(delta);
          }
        }

        if (json.type === 'message_delta' && json.usage) {
          usage = {
            prompt_tokens: 0,
            completion_tokens: json.usage.output_tokens || 0,
            total_tokens: json.usage.output_tokens || 0,
          };
        }

        if (json.type === 'message_start' && json.message?.usage) {
          const u = json.message.usage;
          usage = {
            prompt_tokens: u.input_tokens || 0,
            completion_tokens: 0,
            total_tokens: u.input_tokens || 0,
          };
        }
      } catch {
        // skip
      }
    }
  }

  callbacks.onComplete(fullText, usage);
}

File: src-tauri/src/commands/streaming.rs
use reqwest::Client;
use serde::{Deserialize, Serialize};
use tauri::{AppHandle, Emitter};

#[derive(Deserialize, Debug)]
pub struct StreamRequest {
    pub provider: String,
    pub api_key: String,
    pub model: String,
    pub system_prompt: String,
    pub user_prompt: String,
    pub temperature: Option<f64>,
    pub max_tokens: Option<u32>,
    pub images: Option<Vec<String>>,
    pub stream_id: String,
}

#[derive(Serialize, Clone, Debug)]
pub struct StreamToken {
    pub stream_id: String,
    pub token: String,
    pub done: bool,
    pub error: Option<String>,
    pub usage: Option<StreamUsage>,
}

#[derive(Serialize, Clone, Debug)]
pub struct StreamUsage {
    pub prompt_tokens: u32,
    pub completion_tokens: u32,
    pub total_tokens: u32,
}

#[tauri::command]
pub async fn stream_llm(app: AppHandle, request: StreamRequest) -> Result<(), String> {
    let stream_id = request.stream_id.clone();
    let app_handle = app.clone();

    tokio::spawn(async move {
        let result = match request.provider.as_str() {
            "gemini" => stream_gemini(&app_handle, &request).await,
            "anthropic" => stream_anthropic(&app_handle, &request).await,
            "openai" | "groq" | "openrouter" | "ollama" => {
                stream_openai_compat(&app_handle, &request).await
            }
            _ => Err(format!("Unknown provider: {}", request.provider)),
        };

        if let Err(e) = result {
            let _ = app_handle.emit(
                "llm-stream",
                StreamToken {
                    stream_id: stream_id.clone(),
                    token: String::new(),
                    done: true,
                    error: Some(e),
                    usage: None,
                },
            );
        }
    });

    Ok(())
}

async fn stream_openai_compat(app: &AppHandle, req: &StreamRequest) -> Result<(), String> {
    let base_url = match req.provider.as_str() {
        "openai" => "https://api.openai.com/v1",
        "groq" => "https://api.groq.com/openai/v1",
        "openrouter" => "https://openrouter.ai/api/v1",
        "ollama" => "http://localhost:11434/v1",
        _ => return Err(format!("Unknown provider: {}", req.provider)),
    };
    let url = format!("{}/chat/completions", base_url);

    let mut messages = vec![serde_json::json!({"role": "system", "content": req.system_prompt})];

    if let Some(images) = &req.images {
        if !images.is_empty() {
            let mut content_parts = vec![serde_json::json!({"type": "text", "text": req.user_prompt})];
            for img in images {
                content_parts.push(serde_json::json!({"type": "image_url", "image_url": {"url": img}}));
            }
            messages.push(serde_json::json!({"role": "user", "content": content_parts}));
        } else {
            messages.push(serde_json::json!({"role": "user", "content": req.user_prompt}));
        }
    } else {
        messages.push(serde_json::json!({"role": "user", "content": req.user_prompt}));
    }

    let body = serde_json::json!({
        "model": req.model,
        "messages": messages,
        "temperature": req.temperature.unwrap_or(0.3),
        "max_tokens": req.max_tokens.unwrap_or(8192),
        "stream": true
    });

    let client = Client::new();
    let mut builder = client.post(&url).json(&body);

    if !req.api_key.is_empty() {
        builder = builder.header("Authorization", format!("Bearer {}", req.api_key));
    }
    if req.provider == "openrouter" {
        builder = builder
            .header("HTTP-Referer", "https://zenith-ide.app")
            .header("X-Title", "Zenith IDE");
    }

    let response = builder.send().await.map_err(|e| format!("Network error: {}", e))?;

    if !response.status().is_success() {
        let status = response.status();
        let body = response.text().await.unwrap_or_default();
        return Err(format!("API error {}: {}", status, body));
    }

    let mut stream = response.bytes_stream();
    let mut buffer = String::new();
    use futures::StreamExt;

    while let Some(chunk) = stream.next().await {
        let chunk = chunk.map_err(|e| format!("Stream error: {}", e))?;
        buffer.push_str(&String::from_utf8_lossy(&chunk));

        let lines: Vec<String> = buffer.split('\n').map(|s| s.to_string()).collect();
        buffer = lines.last().cloned().unwrap_or_default();

        for line in &lines[..lines.len().saturating_sub(1)] {
            let trimmed = line.trim();
            if trimmed.is_empty() || trimmed == "data: [DONE]" {
                if trimmed == "data: [DONE]" {
                    let _ = app.emit(
                        "llm-stream",
                        StreamToken {
                            stream_id: req.stream_id.clone(),
                            token: String::new(),
                            done: true,
                            error: None,
                            usage: None,
                        },
                    );
                    return Ok(());
                }
                continue;
            }

            if let Some(data) = trimmed.strip_prefix("data: ") {
                if let Ok(json) = serde_json::from_str::<serde_json::Value>(data) {
                    if let Some(delta) = json["choices"][0]["delta"]["content"].as_str() {
                        let _ = app.emit(
                            "llm-stream",
                            StreamToken {
                                stream_id: req.stream_id.clone(),
                                token: delta.to_string(),
                                done: false,
                                error: None,
                                usage: None,
                            },
                        );
                    }

                    if let Some(usage) = json.get("usage") {
                        let _ = app.emit(
                            "llm-stream",
                            StreamToken {
                                stream_id: req.stream_id.clone(),
                                token: String::new(),
                                done: true,
                                error: None,
                                usage: Some(StreamUsage {
                                    prompt_tokens: usage["prompt_tokens"].as_u64().unwrap_or(0) as u32,
                                    completion_tokens: usage["completion_tokens"].as_u64().unwrap_or(0) as u32,
                                    total_tokens: usage["total_tokens"].as_u64().unwrap_or(0) as u32,
                                }),
                            },
                        );
                        return Ok(());
                    }
                }
            }
        }
    }

    let _ = app.emit(
        "llm-stream",
        StreamToken {
            stream_id: req.stream_id.clone(),
            token: String::new(),
            done: true,
            error: None,
            usage: None,
        },
    );

    Ok(())
}

async fn stream_gemini(app: &AppHandle, req: &StreamRequest) -> Result<(), String> {
    let url = format!(
        "https://generativelanguage.googleapis.com/v1beta/models/{}:streamGenerateContent?alt=sse&key={}",
        req.model, req.api_key
    );

    let mut parts = vec![serde_json::json!({"text": req.user_prompt})];
    if let Some(images) = &req.images {
        for img in images {
            if let Some(captures) = img.strip_prefix("data:") {
                if let Some(idx) = captures.find(";base64,") {
                    let mime = &captures[..idx];
                    let data = &captures[idx + 8..];
                    parts.push(serde_json::json!({"inlineData": {"mimeType": mime, "data": data}}));
                }
            }
        }
    }

    let body = serde_json::json!({
        "system_instruction": {"parts": [{"text": req.system_prompt}]},
        "contents": [{"role": "user", "parts": parts}],
        "generationConfig": {
            "temperature": req.temperature.unwrap_or(0.3),
            "maxOutputTokens": req.max_tokens.unwrap_or(8192)
        }
    });

    let client = Client::new();
    let response = client
        .post(&url)
        .json(&body)
        .send()
        .await
        .map_err(|e| format!("Network error: {}", e))?;

    if !response.status().is_success() {
        let status = response.status();
        let body_text = response.text().await.unwrap_or_default();
        return Err(format!("Gemini API error {}: {}", status, body_text));
    }

    let mut stream = response.bytes_stream();
    let mut buffer = String::new();
    use futures::StreamExt;

    while let Some(chunk) = stream.next().await {
        let chunk = chunk.map_err(|e| format!("Stream error: {}", e))?;
        buffer.push_str(&String::from_utf8_lossy(&chunk));

        let lines: Vec<String> = buffer.split('\n').map(|s| s.to_string()).collect();
        buffer = lines.last().cloned().unwrap_or_default();

        for line in &lines[..lines.len().saturating_sub(1)] {
            let trimmed = line.trim();
            if let Some(data) = trimmed.strip_prefix("data: ") {
                if let Ok(json) = serde_json::from_str::<serde_json::Value>(data) {
                    if let Some(text) = json["candidates"][0]["content"]["parts"][0]["text"].as_str() {
                        let _ = app.emit(
                            "llm-stream",
                            StreamToken {
                                stream_id: req.stream_id.clone(),
                                token: text.to_string(),
                                done: false,
                                error: None,
                                usage: None,
                            },
                        );
                    }

                    if let Some(meta) = json.get("usageMetadata") {
                        let _ = app.emit(
                            "llm-stream",
                            StreamToken {
                                stream_id: req.stream_id.clone(),
                                token: String::new(),
                                done: true,
                                error: None,
                                usage: Some(StreamUsage {
                                    prompt_tokens: meta["promptTokenCount"].as_u64().unwrap_or(0) as u32,
                                    completion_tokens: meta["candidatesTokenCount"].as_u64().unwrap_or(0) as u32,
                                    total_tokens: meta["totalTokenCount"].as_u64().unwrap_or(0) as u32,
                                }),
                            },
                        );
                    }
                }
            }
        }
    }

    let _ = app.emit(
        "llm-stream",
        StreamToken {
            stream_id: req.stream_id.clone(),
            token: String::new(),
            done: true,
            error: None,
            usage: None,
        },
    );

    Ok(())
}

async fn stream_anthropic(app: &AppHandle, req: &StreamRequest) -> Result<(), String> {
    let url = "https://api.anthropic.com/v1/messages";

    let mut content = vec![serde_json::json!({"type": "text", "text": req.user_prompt})];
    if let Some(images) = &req.images {
        for img in images {
            if let Some(captures) = img.strip_prefix("data:") {
                if let Some(idx) = captures.find(";base64,") {
                    let mime = &captures[..idx];
                    let data = &captures[idx + 8..];
                    content.push(serde_json::json!({
                        "type": "image",
                        "source": {"type": "base64", "media_type": mime, "data": data}
                    }));
                }
            }
        }
    }

    let body = serde_json::json!({
        "model": req.model,
        "system": req.system_prompt,
        "messages": [{"role": "user", "content": content}],
        "max_tokens": req.max_tokens.unwrap_or(8192),
        "temperature": req.temperature.unwrap_or(0.3),
        "stream": true
    });

    let client = Client::new();
    let response = client
        .post(url)
        .header("x-api-key", &req.api_key)
        .header("anthropic-version", "2023-06-01")
        .header("Content-Type", "application/json")
        .json(&body)
        .send()
        .await
        .map_err(|e| format!("Network error: {}", e))?;

    if !response.status().is_success() {
        let status = response.status();
        let body_text = response.text().await.unwrap_or_default();
        return Err(format!("Anthropic API error {}: {}", status, body_text));
    }

    let mut stream = response.bytes_stream();
    let mut buffer = String::new();
    use futures::StreamExt;

    while let Some(chunk) = stream.next().await {
        let chunk = chunk.map_err(|e| format!("Stream error: {}", e))?;
        buffer.push_str(&String::from_utf8_lossy(&chunk));

        let lines: Vec<String> = buffer.split('\n').map(|s| s.to_string()).collect();
        buffer = lines.last().cloned().unwrap_or_default();

        for line in &lines[..lines.len().saturating_sub(1)] {
            let trimmed = line.trim();
            if let Some(data) = trimmed.strip_prefix("data: ") {
                if let Ok(json) = serde_json::from_str::<serde_json::Value>(data) {
                    let event_type = json["type"].as_str().unwrap_or("");

                    match event_type {
                        "content_block_delta" => {
                            if let Some(text) = json["delta"]["text"].as_str() {
                                let _ = app.emit(
                                    "llm-stream",
                                    StreamToken {
                                        stream_id: req.stream_id.clone(),
                                        token: text.to_string(),
                                        done: false,
                                        error: None,
                                        usage: None,
                                    },
                                );
                            }
                        }
                        "message_delta" => {
                            if let Some(usage) = json.get("usage") {
                                let _ = app.emit(
                                    "llm-stream",
                                    StreamToken {
                                        stream_id: req.stream_id.clone(),
                                        token: String::new(),
                                        done: true,
                                        error: None,
                                        usage: Some(StreamUsage {
                                            prompt_tokens: 0,
                                            completion_tokens: usage["output_tokens"].as_u64().unwrap_or(0) as u32,
                                            total_tokens: usage["output_tokens"].as_u64().unwrap_or(0) as u32,
                                        }),
                                    },
                                );
                            }
                        }
                        "message_stop" => {
                            let _ = app.emit(
                                "llm-stream",
                                StreamToken {
                                    stream_id: req.stream_id.clone(),
                                    token: String::new(),
                                    done: true,
                                    error: None,
                                    usage: None,
                                },
                            );
                            return Ok(());
                        }
                        _ => {}
                    }
                }
            }
        }
    }

    Ok(())
}
